packages/node-core/src/indexer/dictionary/v2/dictionaryV2.ts

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import {EventEmitter2} from '@nestjs/event-emitter';
+import axios, {AxiosInstance} from 'axios';
+import {DictionaryResponse, DictionaryVersion} from '..';
+import {NodeConfig} from '../../../configure';
+import {BlockHeightMap} from '../../../utils/blockHeightMap';
+import {CoreDictionary} from '../coreDictionary';
+import {subqlFilterBlocksCapabilities} from '../utils';
+import {DictionaryV2Metadata, DictionaryV2QueryEntry} from './types';
+
+export abstract class DictionaryV2<
+  FB,
+  DS,
+  QE extends DictionaryV2QueryEntry = DictionaryV2QueryEntry
+> extends CoreDictionary<DS, FB> {
+  queriesMap?: BlockHeightMap<QE>;
+  protected _metadata: DictionaryV2Metadata | undefined;
+  protected dictionaryApi: AxiosInstance;
+
+  constructor(
+    readonly dictionaryEndpoint: string,
+    protected chainId: string,
+    protected readonly nodeConfig: NodeConfig,
+    protected readonly eventEmitter: EventEmitter2
+  ) {
+    super(dictionaryEndpoint, chainId, nodeConfig, eventEmitter);
+    this.dictionaryApi = axios.create({
+      baseURL: dictionaryEndpoint,
+    });
+    this._dictionaryVersion = DictionaryVersion.v2Complete;
+  }
+
+  protected abstract buildDictionaryQueryEntries(dataSources: DS[]): DictionaryV2QueryEntry;
+
+  async initMetadata(): Promise<void> {
+    this._metadata = await subqlFilterBlocksCapabilities(this.dictionaryEndpoint);
+    this.setDictionaryStartHeight(this._metadata.start);
+  }
+
+  protected get metadata(): DictionaryV2Metadata {
+    if (!this._metadata) {
+      throw new Error(`Fat dictionary _metadata haven't init yet`);
+    }
+    return this._metadata;
+  }
+
+  getQueryEndBlock(startBlockHeight: number, apiFinalizedHeight: number): number {
+    return this.metadata.end;
+  }
+
+  abstract getData(
+    startBlock: number,
+    queryEndBlock: number,
+    limit: number
+  ): Promise<DictionaryResponse<FB> | undefined>;
+
+  queryMapValidByHeight(height: number): boolean {
+    return !!this.queriesMap?.get(height);
```

Comment:
Maybe our map should implement https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/has